### PR TITLE
Add MPI Backend to Enzyme

### PR DESCRIPTION
**Pull request by ludgerpaehler**
The following functions are currently covered with the current backend

- `MPI_Allgather`
- `MPI_Allreduce`
- `MPI_Barrier`
- `MPI_Bcast`
- `MPI_Comm_free`
- `MPI_Comm_rank`
- `MPI_Comm_size`
- `MPI_Comm_split`
- `MPI_Finalize`
- `MPI_Gather`
- `MPI_Scatter`
- `MPI_Get_count`
- `MPI_Graph_create`
- `MPI_Init`
- `MPI_Intercomm_create`
- `MPI_Irecv`
- `MPI_Issend`
- `MPI_Probe`
- `MPI_Recv`
- `MPI_Reduce`
- `MPI_Send`
- `MPI_Test`
- `MPI_Wait`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,3 @@
 *.swp
 *.swo
-build
```
Comment (vchuravy) [inline .gitignore:3]: Unrelated change, please revert

**Review by vchuravy — APPROVED** [review]
Would be good to also add actual runtime tests

**Comment by wsmoses** [comment]
> Would be good to also add actual runtime tests

I don't think runtime tests are as feasible given the weird way mpirun setup may be required. Definitely feasible under certain build conditions (like CI), but they should be a separate target so one doesn't need MPI to build.

Also please add the LLVM tests prior to landing.

**Comment by vchuravy** [comment]
Oh yeah I only want MPI correctness test on CI.

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 2]
```diff
@@ -759,6 +759,15 @@ void calculateUnusedValuesInFunction(
       [&](const Value *val) {
         bool ivn = is_value_needed_in_reverse<ValueType::Primal>(
             TR, gutils, val, mode, PrimalSeen, oldUnreachable);
+        if (!ivn) {
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:762]: Remove this

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 3]
```diff
@@ -937,6 +946,13 @@ void calculateUnusedValuesInFunction(
           return UseReq::Recur;
         bool ivn = is_value_needed_in_reverse<ValueType::Primal>(
             TR, gutils, inst, mode, PrimalSeen, oldUnreachable);
+        if (!ivn) {
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:949]: And this

File: enzyme/Enzyme/GradientUtils.cpp [hunk 4]
```diff
@@ -4756,6 +4763,7 @@ void GradientUtils::computeMinCache(
 
     for (auto V : Intermediates) {
       knownRecomputeHeuristic[V] = !MinReq.count(V);
+      llvm::errs() << "int: " << *V << "minreq: " << MinReq.count(V) << " need: " << NeedGraph.count(V) << "\n";
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:4766]: remove this

File: enzyme/Enzyme/AdjointGenerator.h [hunk 5]
```diff
@@ -5252,7 +6368,7 @@ class AdjointGenerator
       // may load uncacheable data)
       //    Store and reload it
       if (Mode != DerivativeMode::ReverseModeCombined && subretused &&
-          !orig->doesNotAccessMemory()) {
+          (orig->mayWriteToMemory() || !gutils->legalRecompute(orig, ValueToValueMapTy(), nullptr))) {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6371]: restore this to the previous condition

File: enzyme/Enzyme/CacheUtility.cpp [hunk 6]
```diff
@@ -736,6 +736,7 @@ AllocaInst *CacheUtility::createCacheForScope(LimitContext ctx, Type *T,
         }
         Value *es = unwrapM(extraSize, allocationBuilder, available,
                             UnwrapMode::AttemptFullUnwrapWithLookup);
+
```
Comment (wsmoses) [inline enzyme/Enzyme/CacheUtility.cpp:739]: Remove this line add

File: enzyme/test/Enzyme/ReverseMode/context.ll [hunk 7]
```diff
@@ -0,0 +1,143 @@
+; RUN: %opt < %s %loadEnzyme -enzyme -enzyme-preopt=false -mem2reg -instsimplify -early-cse -simplifycfg -correlated-propagation -adce -S | FileCheck %s
```
Comment (wsmoses) [inline enzyme/test/Enzyme/ReverseMode/context.ll:1]: Remove this test/file.

File: enzyme/Enzyme/GradientUtils.cpp (hunk 4 above)
Comment (ludgerpaehler) [reply enzyme/Enzyme/GradientUtils.cpp]: Just removed with the latest commit.

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 8]
```diff
@@ -759,6 +759,13 @@ void calculateUnusedValuesInFunction(
       [&](const Value *val) {
         bool ivn = is_value_needed_in_reverse<ValueType::Primal>(
             TR, gutils, val, mode, PrimalSeen, oldUnreachable);
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:761]: Remove these

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 9]
```diff
@@ -937,6 +944,11 @@ void calculateUnusedValuesInFunction(
           return UseReq::Recur;
         bool ivn = is_value_needed_in_reverse<ValueType::Primal>(
             TR, gutils, inst, mode, PrimalSeen, oldUnreachable);
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:946]: And this
